### ISSUE TITLE
Group and decide which linting rules to include by default.

### ISSUE THREAD

**Issue by ryaneberly** [cflint/CFLint#114]
see #103

**Comment by ryaneberly** [comment]
Proposing we assign the current rules to "groups".  Something like:


Bug Prone
ARG_VAR_CONFLICT
NO_DEFAULT_INSIDE_SWITCH
NESTED_CFOUTPUT
OUTPUT_ATTR
MISSING_VAR

Correctness
ARG_TYPE_MISSING
ARG_TYPE_ANY
ARG_DEFAULT_MISSING
ARG_VAR_MIXED
QUERYNEW_DATATYPE

Bad Practice
GLOBAL_VAR
AVOID_USING_CFDUMP_TAG
AVOID_USING_CFEXECUTE_TAG
AVOID_USING_CFABORT_TAG
AVOID_USING_ABORT
AVOID_USING_WRITEDUMP
AVOID_USING_CFINSERT_TAG
AVOID_USING_CFMODULE_TAG
AVOID_USING_CFUPDATE_TAG

Security
QUERYPARAM_REQ
CFQUERYPARAM_REQ

Code style/Documentation
COMPONENT_HINT_MISSING
FUNCTION_HINT_MISSING-INFO
ARG_HINT_MISSING-INFO
COMPLEX_BOOLEAN_CHECK
EXCESSIVE_FUNCTIONS
FUNCTION_TOO_COMPLEX
EXPLICIT_BOOLEAN_CHECK

Modern Syntax
AVOID_USING_STRUCTNEW
AVOID_USING_ARRAYNEW


And I suggest we include all groups except "Code style/documentation" (and possibly "Modern syntax) by default.

Thoughts?

**Comment by justinmclean** [comment]
Nice grouping. I'd split out the following into a new group as these are less about style and more about maintainability/complexity/lack of abstraction etc:
COMPLEX_BOOLEAN_CHECK
EXCESSIVE_FUNCTIONS
FUNCTION_TOO_COMPLEX
EXPLICIT_BOOLEAN_CHECK

**Comment by justinmclean** [comment]
ARG_TYPE_ANY could possible also go into this new group, but also fine where it is.

**Comment by ryaneberly** [comment]
Good feedback.  I like it.

Complexity
COMPLEX_BOOLEAN_CHECK
EXCESSIVE_FUNCTIONS
FUNCTION_TOO_COMPLEX
EXPLICIT_BOOLEAN_CHECK

**Comment by justinmclean** [comment]
Also add to same group:
EXCESSIVE_FUNCTION_LENGTH
EXCESSIVE_COMPONENT_LENGTH

**Comment by justinmclean** [comment]
also add (and there's more to come):

Poor Naming:
VAR_INVALID_NAME
VAR_ALLCAPS_NAME
VAR_TOO_SHORT
VAR_TOO_LONG
VAR_TOO_WORDY
VAR_IS_TEMPORARY
VAR_HAS_PREFIX_OR_POSTFIX
METHOD_INVALID_NAME
METHOD_TOO_SHORT
METHOD_TOO_LONG
METHOD_TOO_WORDY
METHOD_IS_TEMPORARY

**Comment by justinmclean** [comment]
Also in code style add:
FUNCTION_TYPE_MISSING
FUNCTION_TYPE_ANY

**Comment by justinmclean** [comment]
In bad practice add:
LOCAL_LITERAL_VALUE_USED_TOO_OFTEN
GLOBAL_LITERAL_VALUE_USED_TOO_OFTEN

**Comment by justinmclean** [comment]
Also add to complexity:
EXCESSIVE_ARGUMENTS

**Comment by ryaneberly** [comment]
with justinmclean's suggestions

Bug Prone
ARG_VAR_CONFLICT
NO_DEFAULT_INSIDE_SWITCH
NESTED_CFOUTPUT
OUTPUT_ATTR
MISSING_VAR

Correctness
ARG_DEFAULT_MISSING
ARG_TYPE_ANY
ARG_TYPE_MISSING
ARG_VAR_MIXED
QUERYNEW_DATATYPE

Bad Practice
AVOID_USING_ABORT
AVOID_USING_CFABORT_TAG
AVOID_USING_CFDUMP_TAG
AVOID_USING_CFEXECUTE_TAG
AVOID_USING_CFINSERT_TAG
AVOID_USING_CFMODULE_TAG
AVOID_USING_CFUPDATE_TAG
AVOID_USING_WRITEDUMP
GLOBAL_LITERAL_VALUE_USED_TOO_OFTEN
GLOBAL_VAR
LOCAL_LITERAL_VALUE_USED_TOO_OFTEN

Security
CFQUERYPARAM_REQ
QUERYPARAM_REQ

Code style/Documentation
ARG_HINT_MISSING-INFO
COMPONENT_HINT_MISSING
FUNCTION_HINT_MISSING-INFO
FUNCTION_TYPE_ANY
FUNCTION_TYPE_MISSING

Modern Syntax
AVOID_USING_ARRAYNEW
AVOID_USING_STRUCTNEW

Complexity
COMPLEX_BOOLEAN_CHECK
EXCESSIVE_FUNCTIONS
EXCESSIVE_ARGUMENTS
EXPLICIT_BOOLEAN_CHECK
EXCESSIVE_COMPONENT_LENGTH
EXCESSIVE_FUNCTION_LENGTH
FUNCTION_TOO_COMPLEX

Poor Naming:
METHOD_HAS_PREFIX_OR_POSTFIX
METHOD_INVALID_NAME
METHOD_IS_TEMPORARY
METHOD_TOO_SHORT
METHOD_TOO_LONG
METHOD_TOO_WORDY
VAR_ALLCAPS_NAME
VAR_HAS_PREFIX_OR_POSTFIX
VAR_INVALID_NAME
VAR_IS_TEMPORARY
VAR_TOO_SHORT
VAR_TOO_LONG
VAR_TOO_WORDY

**Comment by denuno** [comment]
Love'n the groups!  I just added a GUI for CFLint to CFEclipse, and these groups would help organize it better.  To make editing/keeping track easier, I created a wiki page for the group list here: https://github.com/cflint/CFLint/wiki/Proposed-Rule-Groups

Would the groups define the rules, or would we define the groups in the rules?  Eg:

`"security" : ["CFQUERYPARAM_REQ","QUERYPARAM_REQ"]`
versus
```
    {
      "name": "ArgVarChecker",
      "message": [
        {
          "code": "ARG_VAR_CONFLICT",
          "messageText": "Variable ${variable} should not be declared in both var and argument scopes.",
          "severity": "ERROR",
          "groups":["buggy"]
        },
```

Would it be one group per rule, or could some rules be in more than one group?

I think I like rules being able to be in more than one group, but it might be better if that's not allowed...
